README.md
bash-shellshock wrapper
=======================

This is a small wrapper around /bin/bash that refuses to start bash if any
environment variables start with '('.  It can also be run in a log-only mode
and a mode that strips these 'bad' environment variables.

You can install this as a temporary workaround if you don't fully trust the
latest patches for CVE-2014-6271 and CVE-2014-7169, which (as of 2014-09-25)
seem like they've been a bit rushed.


Example session
---------------

```sh
$ ls -l /bin/bash*
lrwxrwxrwx 1 root root      20 Sep 26 01:12 /bin/bash -> /bin/bash-shellshock
-rwxr-xr-x 1 root root 1029624 Sep 24 11:51 /bin/bash.real
-rwxr-xr-x 1 root root   10368 Sep 26 00:32 /bin/bash-shellshock

$ XX=1 XXX='()hello' /bin/bash -c env
bash-shellshock: Refusing to start due to possibly unsafe environment variable (see syslog)

$ XX=1 XXX='([world' /bin/bash -c env
bash-shellshock: Refusing to start due to possibly unsafe environment variable (see syslog)

$ sudo touch /etc/bash-shellshock.strip-vars

$ XX=1 XXX='()hello' bash -c env | grep ^XX
XX=1

$ sudo touch /etc/bash-shellshock.log-only

$ XX=1 XXX='()hello' bash -c env | grep ^XX
XX=1
XXX=()hello

$ sudo grep 'bash-shellshock\[' /var/log/auth.log
Sep 26 01:51:31 debian bash-shellshock[5003]: (PPID=4708 UID=1000 GID=1000 EUID=1000 EGID=1000 CWD=/home/dwon) \
    Refusing to start due to possibly unsafe environment variable: XXX=()hello
Sep 26 01:51:36 debian bash-shellshock[5009]: (PPID=4708 UID=1000 GID=1000 EUID=1000 EGID=1000 CWD=/home/dwon) \
    Refusing to start due to possibly unsafe environment variable: XXX=([world
Sep 26 01:51:41 debian bash-shellshock[5018]: (PPID=4708 UID=1000 GID=1000 EUID=1000 EGID=1000 CWD=/home/dwon) \
    Stripping possibly unsafe environment variable: XXX=()hello
Sep 26 01:51:50 debian bash-shellshock[5034]: (PPID=4708 UID=1000 GID=1000 EUID=1000 EGID=1000 CWD=/home/dwon) \
    Possibly unsafe environment variable: XXX=()hello

$ rm -f /etc/bash-shellshock.log-only /etc/bash-shellshock.strip-vars
```


Downloading
-----------

- Binary .deb packages for Debian and Ubuntu:

    https://github.com/dlitz/bash-shellshock/tree/binaries/binaries

- Source code:

    https://github.com/dlitz/bash-shellshock


Building from source
--------------------

* To build and install this package from source under Debian or Ubuntu, clone
  the repository and run:

```sh
dpkg-buildpackage -us -uc -tc
sudo dpkg -i ../bash-shellshock_1.0_amd64.deb
```

  See also the [docs/sbuild.md](docs/sbuild.md) file for information about
  building using sbuild.

* To compile manually, run this:

```sh
make
make install
ln /bin/bash /bin/bash.real
ln -sf /bin/bash-shellshock /bin/bash
```

  You may need to adjust the paths in the Makefile.


Configuration
-------------

If bash-shellshock finds environment variables starting with '(', it will do
one of three things (in order of priority):

1. If `/etc/bash-shellshock.log-only` exists, it will log the event to syslog
   and exec `/bin/bash.real`.

2. If `/etc/bash-shellshock.strip-vars` exists, it will log the event to
   syslog, remove the variable from the environment, and exec `/bin/bash.real`.

3. Otherwise, it will log the event to syslog and abort with exit status 255.


Testing
-------

* For CVE-2014-6271

```sh
env X='() { ignored; }; /bin/echo "You are vulnerable to CVE 2014-6271"' bash -c true
```

* For CVE-2014-7169 ([reference](https://twitter.com/taviso/status/514887394294652929)):

```sh
env X='() { (a)=>\' bash -c "echo date"; cat echo
```


Logging
-------

Logs go to the `LOG_AUTHPRIV` syslog facility, which is typically found at `/var/log/auth.log`.


Caveat
------

This might break bash scripts that rely on receiving functions and/or arrays
via the environment.  You probably shouldn't do that anyway.


Better solution
---------------

Remove this silly feature from bash in the first place.


License
-------
Copyright 2014 InboxApp Inc.

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the "Software"),
to deal in the Software without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of the Software, and to permit persons to whom the
Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


debian/changelog
bash-shellshock (1.1) unstable; urgency=medium

  * Reduce the scope of the check to variables that start with '()' or '(['.
    This unbreaks virtualenv and other tools that set PS1='(foo) ...'.

 -- Dwayne Litzenberger <dlitz@dlitz.net>  Thu, 09 Oct 2014 15:33:21 -0700

bash-shellshock (1.0) unstable; urgency=low

  * Fix strip-vars mode and add unit tests.

 -- Dwayne Litzenberger <dlitz@dlitz.net>  Thu, 25 Sep 2014 23:59:47 -0700

bash-shellshock (0.0) unstable; urgency=low

  * Initial Release.

 -- Dwayne Litzenberger <dlitz@dlitz.net>  Thu, 25 Sep 2014 18:00:19 -0700


testenv.c
/*
 * Usage:
 *  ./testenv.t
 *      - Print the environment
 *  ./testenv.t FOO=BAR... -- CMD [ARG...]
 */
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
extern char **environ;

int main(int argc, char **argv) {
    if (argc == 1) {
        char **p;
        for (p = environ; *p != NULL; p++) {
            puts(*p);
        }
        exit(0);
    } else {
        char **env = argv + 1;
        for (; *argv != NULL; argv++) {
            if (argv[0][0] == '-' && argv[0][1] == '-' && argv[0][2] == '\0') {
                *argv = NULL;
                argv++;
                break;
            }
        }
        if (*argv == NULL) {
            fprintf(stderr, "testenv: No argv specified\n");
            exit(127);
        }

        execve(argv[0], argv, env);
        perror("testenv: command failed");
        exit(127);
    }
}


bash-shellshock.c
/* Wrapper for bash -- removes environment variables values start with "("
 * 2014-09-25 Dwayne Litzenberger <dlitz@dlitz.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included
 * in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
 * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>
extern char **environ;

#ifndef PROG_NAME
#define PROG_NAME "bash-shellshock"
#endif

#ifndef REAL_BASH
#define REAL_BASH "/bin/bash.real"
#endif

#ifndef CONFIG_FILE_LOG_ONLY
#define CONFIG_FILE_LOG_ONLY "/etc/bash-shellshock.log-only"
#endif

#ifndef CONFIG_FILE_STRIP_VARS
#define CONFIG_FILE_STRIP_VARS "/etc/bash-shellshock.strip-vars"
#endif

static char cwdbuf[8192];
static char *cwd;
static pid_t ppid;
static uid_t uid, euid;
static gid_t gid, egid;
static char *info;  /* for logging */

#define MODE_UNINITIALIZED 0
#define MODE_HARD_EXIT 1
#define MODE_STRIP_VARS 2
#define MODE_LOG_ONLY 3

static int config_mode = MODE_UNINITIALIZED;

static void setup(void) {
    struct stat st;

    /* Load configuration */
    if (0 == lstat(CONFIG_FILE_LOG_ONLY, &st)) {
        config_mode = MODE_LOG_ONLY;
    } else if (0 == lstat(CONFIG_FILE_STRIP_VARS, &st)) {
        config_mode = MODE_STRIP_VARS;
    } else {
        config_mode = MODE_HARD_EXIT;
    }

    /* Get a bunch of info for logging */
    ppid = getppid();
    uid = getuid();
    euid = geteuid();
    gid = getgid();
    egid = getegid();
    cwd = getcwd(cwdbuf, sizeof(cwdbuf) - 1);
    if (NULL == cwd) {
        cwd = "";
    }

    if (asprintf(&info, "PPID=%d UID=%d GID=%d EUID=%d EGID=%d CWD=%s", ppid, uid, gid, euid, egid, cwd) < 0) {
        info = "<err>";
    }

    openlog(PROG_NAME, LOG_PID | LOG_CONS, LOG_AUTHPRIV);
}

int main(int argc, char **argv) {
    char **src;
    char **dest;
    char *eq;

    for(src = dest = environ; *src != NULL; src++, dest++) {
        *dest = *src;   /* no-op except when MODE_STRIP_VARS is in use */

        /* Detect special environment variables:
         *  - starting with "()" -- possible shell function
         *  - starting with "([" -- possible array variable
         * See bash's variables.c file.
         *
         * Previously, we just looked for an initial '(', but that broke things
         * like virtualenv's "activate" script, which sets PS1="(foo) $PS1".
         */
        eq = strchr(*src, '=');
        if (eq != NULL && eq[1] == '(' && (eq[2] == ')' || eq[2] == '[')) {
            if (config_mode == MODE_UNINITIALIZED) {
                setup();
            }

            if (config_mode == MODE_LOG_ONLY) {
                syslog(LOG_WARNING, "(%s) Possibly unsafe environment variable: %s", info, *src);
            } else if (config_mode == MODE_STRIP_VARS) {
                syslog(LOG_WARNING, "(%s) Stripping possibly unsafe environment variable: %s", info, *src);
                dest--;
            } else {
                syslog(LOG_CRIT, "(%s) Refusing to start due to possibly unsafe environment variable: %s", info, *src);
                fprintf(stderr, "bash-shellshock: Refusing to start due to possibly unsafe environment variable (see syslog)\n");
                exit(255);
            }
        }
    }
    *dest = NULL;

    execve(REAL_BASH, argv, environ);
    syslog(LOG_ERR, "(%s) Failed to start " REAL_BASH ": errno=%d", info, errno);
    exit(127);
}
